bitcoin/src/address/mod.rs

```diff
@@ -665,12 +665,18 @@ impl Address<NetworkUnchecked> {
     ///
     /// For details about this mechanism, see section [*Parsing addresses*](Address#parsing-addresses)
     /// on [`Address`].
+    ///
+    /// # Errors
+    ///
+    /// This function only ever returns the [`ParseError::NetworkValidation`] variant of
+    /// `ParseError`. This is not how we normally implement errors in this library but
+    /// `require_network` is not a typical function, it is conceptually part of string parsing.
     #[inline]
-    pub fn require_network(self, required: Network) -> Result<Address, NetworkValidationError> {
+    pub fn require_network(self, required: Network) -> Result<Address, ParseError> {
```

Comment:
I don't think we need to change this to satisfy #2507. The into should take care of it. 
```rust
let address = s.parse::<Address<_>>()?
    .require_network(network)?;
```

Reply:
@Kixunil explicitly wants an error that is returnable without users having to create one, that's why this PR returns `ParseError`. This is one of those PRs that has loads of context spread over various PRs and issues and was discussed heavily by myself, @apoelstra, and Kix. FTR I personally do not like the solution but have not managed to come up with a better one. I did try in https://github.com/rust-bitcoin/rust-bitcoin/pull/2593

Reply:
I can see how this is annoying

Reply:
#2593 was maaybe a cleaner solution because it gives the user a new `address::Error` they can return using `From`, but the issue there is that if they want to further wrap `address::Error` then suddenly `?` will stop working again.

Also, it's not written in #2507 but I would also like `s.parse().and_then(|a| a.require_network(network))` to work, and there are no hidden `from`s in there.

Ultimately the real problem here is that Rust's monadic error handling is not that great. I agree that this solution isn't great but I think it's the best thing.

Reply:
I believe we will face a similar problem with other error changes too. Whenever there is a 3 level hierarchy of errors, we will have this issue. 

Can we flatten all errors and having a single `bitcoin::Error` and other child errors.

Reply:
I haven't hit any other problems as yet and I've tested all the crates upto and including `bdk`. Also we have these nested errors already in `v0.31.0` with no problems. I think address parsing is a special case.

Reply:
I too would prefer to return `NetworkValidationError` here but it's not a big deal and I don't want to stall release because of it.

Reply:
> I too would prefer to return `NetworkValidationError` here but it's not a big deal and I don't want to stall release because of it.

I"m confused. You opened #2507 saying it was a "shitstorm" to return `NetworkValidationError` here.

Reply:
Oh, crap, I missed this. Shitstorm was not having an error you can `?` to. I was expecting `ParseError: From<NetworkValidationError>` but return `NetworkValidationError` here. However you make a good point that using `?` with user-defined type becomes more annoying since the user needs to impl `From` for both types and also `and_then` becomes more annoying as well.

Honestly I don't like either approach so for the sake of not changing the API all the time I'm not going to reopen this.

Reply:
Feel free to open an issue in the future if you want to discuss an alternate solution (further discussion here will definitely get lost). But unfortunately I think you're right that there's no "good" and ergonomic solution.